### PR TITLE
359 Regression tree-based AL

### LINKED ISSUE #359: Implementation of RT-AL
Implementation [Regression tree‑based active learning](https://link.springer.com/article/10.1007/s10618-023-00951-7) (RT-AL)

### PR DESCRIPTION
**Pull request by Pascal112**
Closes #359

### CODE REVIEW

**Review by mherde — CHANGES_REQUESTED** [review]
Further discussion will follow in the next meeting.

File: skactiveml/pool/tests/test_regression_tree_based_al.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    def test_query(self):
+        qs = self.qs_class(random_state=0)
+        X = np.array([0, 2, 10, 12, 20, 22, 1, 11, 21]).reshape(-1, 1)
+        y = np.append([0, 2, 10, 12, 20, 22], np.full(3, MISSING_LABEL))
+        batch_size = 3
+
+        idxs, utilities = qs.query(
+            X, y, self.reg, batch_size=batch_size, return_utilities=True)
+        self.reg.fit(X, y)
+        np.testing.assert_array_equal(np.nansum(utilities[0]), batch_size)
```
Comment (mherde) [inline skactiveml/pool/tests/test_regression_tree_based_al.py:82]:
```suggestion
```
Test is redundant due to the subsequent line of code.

File: skactiveml/pool/tests/test_regression_tree_based_al.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+            def apply(self, X):
+                return np.argmin(
+                    abs(
+                        np.tile(X, (len(self.centers)))
+                        -np.tile(self.centers, (len(X), 1))),
+                    axis=1
+                )
+
+        DELTA = np.array([1, 1, 1])
```
Comment (mherde) [inline skactiveml/pool/tests/test_regression_tree_based_al.py:99]:
```suggestion
        delta = np.array([1, 1, 1])
```

File: docs/examples/pool/_regression_tree_based_al.json [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    "query_params" : "X=X, y=y, reg=reg, return_utilities=True",
+    "batch_size" : "3"
+  },
+  {
+    "class" : "RegressionTreeBasedAL",
+    "package" : "pool",
+    "method" : "Regression Tree Based Active Learning (RT-AL) with Representativity Selection",
+    "template" : "examples/template_pool_regression_2.py",
+    "tags" : ["pool",  "regression", "single-annotator"],
+    "title" : "Regression Tree Based Active Learning (RT-AL) with representativity Selection",
```
Comment (mherde) [inline docs/examples/pool/_regression_tree_based_al.json:42]:
```suggestion
    "title" : "Regression Tree Based Active Learning (RT-AL) with Representativity Selection",
```

**Review by mherde — CHANGES_REQUESTED** [review]

**Review by mherde — APPROVED** [review]

File: docs/examples/template_pool_regression_batch.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+noise = np.vectorize(lambda x : random_state.rand() * 1.5 if x < 1
+                                else random_state.rand() * 0.5)
+
+# Build a dataset.
+y_true = true_function(X) + noise(X).flatten()
+y = np.full(shape=y_true.shape, fill_value=MISSING_LABEL)
+X_test = np.linspace(0, 2, num="$res|1000").reshape(-1, 1)
+
+# Initialise the classifier.
```
Comment (AshnaJose) [inline docs/examples/template_pool_regression_batch.py:34]: regressor

File: skactiveml/pool/_regression_tree_based_al.py [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+                    )  # Equation (10)
+
+                    # Use Eq. (8) to find the next sample to be labeled.
+                    R_l = R_cand[l_cand == l]
+                    batch_utilities_cand[l, l_cand == l] = delta_l - R_l
+                    best_indices[l] = rand_argmax(
+                        batch_utilities_cand[l],
+                        random_state=self.random_state_,
+                    )
+
```
Comment (mherde) [inline skactiveml/pool/_regression_tree_based_al.py:281]: Can we use the `simple_batch ` at the end for representativity?

**Review by mherde — CHANGES_REQUESTED** [review]

**Review by mherde — APPROVED** [review]

**Pull request merged**